ClassAttendance/src/BLL/Services/GroupService.cs

```diff
@@ -0,0 +1,114 @@
+﻿using AutoMapper;
+using BLL.Interfaces;
+using BLL.Models;
+using DAL.DTO;
+using DAL.Interfaces;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace BLL.Services
+{
+    public class GroupService : IGroupService
+    {
+        private readonly IStore<GroupDto> _groups;
+        private readonly IMapper _mapper;
+
+        public async Task<int> AddAsync(Group item)
+        {
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            if (string.IsNullOrEmpty(item.Name))
+            {
+                throw new ArgumentException("Group name can't be empty");
+            }
+
+            var dto = _mapper.Map<GroupDto>(item);
+            await _groups.AddAsync(dto);
+            return dto.Id;
+        }
+
+        public async Task DeleteAsync(int id)
+        {
+            await _groups.DeleteAsync(id);
+        }
+
+        public IAsyncEnumerable<Group> GetAll()
+        {
+            var dtos = _groups.GetAll().AsAsyncEnumerable();
+            var models = _mapper.Map<IAsyncEnumerable<Group>>(dtos);
```

Comment:
As i know, MS recomend to use IAsyncEnumerable instead of Task<IEnumerable>. I will change this to IEnumerable if needed.

https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8

https://dotnetcoretutorials.com/2019/01/09/iasyncenumerable-in-c-8/

Reply:
Yes, I agreed that AsAsyncEnumerable is a good thing, but I think the collection that you got from AutoMapper should be IEnumerable because AutoMapper read all entities from source collection in synchronous way.

Reply:
Fixed